source/common/grpc/google_async_client_impl.cc

```diff
@@ -10,18 +10,56 @@
 namespace Envoy {
 namespace Grpc {
 
+GoogleAsyncClientThreadLocal::GoogleAsyncClientThreadLocal()
+    : completion_thread_(new Thread::Thread([this] { completionThread(); })) {}
+
+GoogleAsyncClientThreadLocal::~GoogleAsyncClientThreadLocal() {
+  // Force streams to shutdown and invoke TryCancel() to start the drain of
+  // pending op. If we don't do this, Shutdown() below can jam on pending ops.
+  // This is also required to satisfy the contract that once Shutdown is called,
+  // streams no longer queue any additional tags.
+  for (auto it = streams_.begin(); it != streams_.end();) {
+    (*it++)->resetStream();
+  }
+  cq_.Shutdown();
+  ENVOY_LOG(debug, "Joining completionThread");
+  completion_thread_->join();
+  ENVOY_LOG(debug, "Joined completionThread");
+  // Ensure that we have cleaned up all orphan streams, now that CQ is gone.
+  while (!streams_.empty()) {
+    (*streams_.begin())->onCompletedOp();
+  }
+}
+
+void GoogleAsyncClientThreadLocal::completionThread() {
+  ENVOY_LOG(debug, "completionThread running");
+  void* tag;
+  bool ok;
+  while (cq_.Next(&tag, &ok)) {
+    const auto& google_async_tag = *reinterpret_cast<GoogleAsyncTag*>(tag);
+    const GoogleAsyncTag::Operation op = google_async_tag.op_;
+    GoogleAsyncStreamImpl& stream = google_async_tag.stream_;
+    ENVOY_LOG(trace, "completionThread CQ event {} {}", op, ok);
+    std::unique_lock<std::mutex> lock(stream.completed_ops_lock_);
+    stream.completed_ops_.emplace_back(op, ok);
+    stream.dispatcher_.post([&stream] { stream.onCompletedOp(); });
```

Comment:
For perf reasons, you might want to only post if the number of completed ops is going from 0 -> 1, though from below it seems like you only want to process 1 at a time. I'm not exactly sure why that is, but might be worth throwing in a TODO here for future perf follow ups. (You could just keep track of number of per silo pending events, and post into that queue at the thread level).

Reply:
This was actually the original design. Unfortunately, if you batch process `completed_ops_`, rather than have a 1:1 `onCompleteOp` association with `completed_ops_`, you get into a race, where the streams may have processed the batched ops and then destroyed itself (when inflight ops hits zero), and then a dangling dispatcher post is executed and tries to access the freed stream.

Reply:
You can't wait on a counter of number of pending `post`s either, since this won't work in the global teardown case that we discussed yesterday.

Reply:
I think what's written here is safe, while not the most efficient. I'd prefer a TODO now rather than complicating this PR further. It would be easier to reason about the improved behavior separately.